Improve dmx binding

J-N-K opened the pull request:
This will add tunable white to the dimmer thing handler, improve brightness handling for RGB lights and include improved tests.

- [x] improve binding
- [x] update documentation
- [x] testing

J-N-K commented:
retriggering build due to p2 repository failure

maggu2810 commented:
If it fails or not fails which log severity is used?
If e.g. the logger is using debug and the code is using trace, then the logging is not done. If the code is using debug, it is done. It could be timing related.
Could you e.g. set the log severity to trace and check if it will fail then too

J-N-K commented:
That is exactly what is happening. Removing the logging (or setting to trace) leads to test failures. But I have no idea how to fix the timing in the tests. As far as I can tell, this works flawlessly in production.

J-N-K commented:
How can i set the log level during the tests in a build? I see that it is set to DEBUG, but I can't find the setting.

maggu2810 commented:
I have to admit that I don't know how we configure the logging at the moment on test execution without a look at the sources.

But it should be easy to inject a slf4j logging implementation (e.g. log4j or logback backend) into the surefire test execution to support configurable logging for development.

J-N-K commented:
Unfortunately, even after days of research, I was not able to find out how to do that. Logback is a requirement for the tests, but I don't have the slightest idea how to configure that.

maggu2810 commented:
@J-N-K I will try to have a look at (some time).

maggu2810 commented:
@J-N-K For me it has been enough to set the system property for the logback configuration path on the test execution. logback itself is already present in the "current" test runtime configuration.
See: https://github.com/eclipse/smarthome/commit/9392d342b85b580484c6bc83580a6355f05e7e93

J-N-K commented:
Thanks. Works like a charm.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.dmx.test/src/test/java/org/eclipse/smarthome/binding/dmx/ColorThingHandlerTest.java:
```diff
@@ -0,0 +1,319 @@
+    public void setUp() {
+        registerService(volatileStorageService);
+        managedThingProvider = getService(ThingProvider.class, ManagedThingProvider.class);
+        assertThat("Could not get ManagedThingProvider", managedThingProvider, is(notNullValue()));
```
See above, please remove the string comments.

J-N-K commented on extensions/binding/org.eclipse.smarthome.binding.dmx/src/main/java/org/eclipse/smarthome/binding/dmx/handler/ColorThingHandler.java:
```diff
@@ -0,0 +1,300 @@
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "turn-on value malformed");
+                dmxHandlerStatus = ThingStatusDetail.CONFIGURATION_ERROR;
+                return;
```
Can this really happen? `thingUpdated` and `handleConfigurationUpdate`both call `dispose()` before`ìnitialize()`. Anyway, I can move the listener registration some lines down (e.g. l. 259). That way the listeners will only be registered if the configuration is ok. Would that be ok with you?

triller-telekom replied:
Yes that will be much better, thank you.

J-N-K commented on extensions/binding/org.eclipse.smarthome.binding.dmx/src/main/java/org/eclipse/smarthome/binding/dmx/handler/ColorThingHandler.java:
```diff
@@ -0,0 +1,300 @@
+            case CHANNEL_BRIGHTNESS_R:
+                if (command instanceof RefreshType) {
+                    logger.trace("sending update on refresh to channel {}:brightness_r", this.thing.getUID());
+                    updateState(channelUID, Util.toPercentValue(currentValues.get(0)));
```
Maybe there is a misunderstanding on my side regarding the `Refresh` command. What exactly is supposed to happen? From my understanding, the bidning should update the channel state so that it reflects the device state. Is that wrong? 
Regarding the `currentValue`: There is no real status feedback as DMX is one-way communication. The binding keeps track of the (assumed) value of each DMX channel in the universe. So I might get that information from the universe (which is the same as in `currentValue` if the binding is working as it should).

triller-telekom replied:
Ah, Ok now I see. So the misunderstanding was on my side. You are right, the `refresh` command is issued by the framework to query the current state of the real physical device. So if you only have a one way communication, returning the one from your universe is the right thing to do.

J-N-K commented on extensions/binding/org.eclipse.smarthome.binding.dmx/src/main/java/org/eclipse/smarthome/binding/dmx/handler/ColorThingHandler.java:
```diff
@@ -0,0 +1,300 @@
+                return;
+        }
+        currentColor = HSBType.fromRGB(currentValues.get(0), currentValues.get(1), currentValues.get(2));
+        updateState(new ChannelUID(this.thing.getUID(), CHANNEL_COLOR), currentColor);
```
The `COLOR`  channel is high level channel for convient use of the thing. Its value is calculated from the lower level channels (i.e. the R,G,B components). 
The feedback is only provided on a physical channel (i.e.  R,G,B), but the user wants to see the change in the `COLOR`, so on receiving one feedback, both need to be updated. This could only be avoided if the low-level channels are not exposed to the user, but IMO this would impose a limit on functionality for the sake of "nicer" code. 
The same is true for the `TunableWhiteThingHandler` below: the physical channels are cold white and warm white, the total brightness ond the color temperature is only calculated from those.

J-N-K commented:
I incorporated all feedback and rebased on master to include the mockito2 and SAT upgrades. The only open question is the upgrade of multiple channels. Since my comment is hard to find, I'll repost it here:

The `COLOR` channel in the `ColorThingHandler` is high level channel for convient use of the thing. Its value is calculated from the lower level channels (i.e. the R,G,B components).
The feedback is only provided on a physical channel (i.e. R,G,B), but the user wants to see the change in the `COLOR`, so on receiving one feedback, both need to be updated. This could only be avoided if the low-level channels are not exposed to the user, but IMO this would impose a limit on functionality for the sake of "nicer" code.
The same is true for the `TunableWhiteThingHandler` below: the physical channels are the brightness of cold white and warm white, the total brightness ond the color temperature is only calculated from those.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.dmx.test/src/test/java/org/eclipse/smarthome/binding/dmx/test/TestBridgeHandler.java:
```diff
@@ -88,6 +88,12 @@ public void initialize() {
         universe.calculateBuffer(time + timespan);
+        try {
+            // needed for the framework to process all events
+            Thread.sleep(200);
```
This will make the tests rather unstable. Please use a `waitForAssert` call.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.dmx.test/src/test/java/org/eclipse/smarthome/binding/dmx/test/TestBridgeHandler.java:
```diff
@@ -88,12 +88,7 @@ public void initialize() {
-        } catch (InterruptedException e) {
-            logger.error("failed to sleep for 200 ms");
-        }
+        logger.debug("calculated buffer for {} to {} (difference {})", time, time + timespan, timespan);
```
Now you have just removed the waiting time but before you argued that the framework needs time to process your events.

So you probably have to wait until the event is fully processed at the place where you call `calcBuffer`. Am I right?

You can find an example for the `waifForAssert` [here](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/items/GroupItemTest.java#L104)

J-N-K replied:
This was a leftover from an earlier attempt to fix the randomly failing tests. After some more investigation, this real problem is that sometimes the updates from the things seem to disappear. SInstead a waitforAssert is implemented in all tests, waiting for the thing channel update.

J-N-K commented:
Is there anything I need to do here?
